### PR TITLE
rules: Disable screen rotation for chassis_type other than  30, 31, a…

### PR DESCRIPTION
**Pull request by KaiChuan-Hsieh**
…nd 32

For product's chassis_type is not {30, 31, 32},
disable rotation on all Dell product with Intel sensor_hub embedded.

The chassis_type info:
30 -> tablet
31 -> convertible
32 -> detachable

### CODE REVIEW

**Comment by KaiChuan-Hsieh** [comment]
May I know if ATTR{[dmi/id]chassis_type} is not a good way to retrieve chassis_type attribute? The autopkgtest is failed, the rule-syntax-check.py does not allow the pattern. I'd like to know if there is any better way to retrieve the chassis_type since I need it to determine if the hardware can disable rotation.

**Comment by keszybz** [comment]
It doesn't like `ATTR{[dmi/id]chassis_type}`. The regexp only allows `a-zA-Z0-9/_.*%-`. Two problems here: most likely the regexp should be adjusted, but also the syntax with `[]` should be documented in the man page.

**Comment by KaiChuan-Hsieh** [comment]
@keszybz 
Thanks for your reply. Actually, I can't find the detail for the usage. I browse the website and find the usage for retrieving dmi/id attribute, and it works for me.
https://lwn.net/Articles/456217/

Is there any quick way for me to pick up it? or I can just go through the udevd source for knowing how it process the string?

File: hwdb.d/60-sensor.hwdb [hunk 1]
```diff
@@ -262,7 +262,7 @@ sensor:modalias:acpi:*KIOX000A*:dmi:*svn*CytrixTechnology:*pn*Complex11t*
 #########################################
 # Dell
 #########################################
-sensor:modalias:platform:HID-SENSOR-200073:dmi:*svnDell*:pnVostro5581:*
+sensor:modalias:platform:HID-SENSOR-200073:dmi:*svnDell*
```
Comment (keszybz) [inline hwdb.d/60-sensor.hwdb:265]: Can we be sure that this sensor is only used in the base (for all Dell products)?

Comment (KaiChuan-Hsieh) [reply hwdb.d/60-sensor.hwdb:265]: Sure, I'll request Dell's PM to confirm it.

**Comment by khfeng** [comment]
DMI chassis types are quite often incorrectly filled. So I believe we should stick to current model, i.e. each model has its own quirk entry.

@jwrdegoede may have more input on this.

**Comment by superm1** [comment]
> DMI chassis types are quite often incorrectly filled. So I believe we should stick to current model, i.e. each model has its own quirk entry.
> 
> @jwrdegoede may have more input on this.

On Dell's systems you've seen it wrong?  Any details would be really useful.  To be honest - I'd be really surprised it's wrong on any major OEM - particularly on systems shipping with Windows 10.  The specific requirements for https://docs.microsoft.com/en-us/windows-hardware/test/hlk/testref/751627d6-dde1-47d9-81a4-057a2aa1a3be prohibit it being wrong.

File: rules.d/60-sensor.rules [hunk 2]
```diff
@@ -3,6 +3,10 @@
 ACTION=="remove", GOTO="sensor_end"
 
 # device matching the sensor's name and the machine's DMI data for IIO devices
+SUBSYSTEM=="iio", KERNEL=="iio*", SUBSYSTEMS=="usb|i2c|platform", \
+  KERNELS=="HID-SENSOR-200073*", ATTR{[dmi/id]chassis_type}=="30|31|32", \
```
Comment (superm1) [inline rules.d/60-sensor.rules:7]: This makes sense to me.

File: hwdb.d/60-sensor.hwdb (hunk 1 above)
Comment (KaiChuan-Hsieh) [reply hwdb.d/60-sensor.hwdb:265]: @keszybz 

with @superm1 confirmation, I think the rule can apply to all Dell platforms, please let me know if you have other concern.

Thanks,

**Comment by yuwata** [comment]
> It doesn't like ATTR{[dmi/id]chassis_type}. The regexp only allows a-zA-Z0-9/_.*%-. Two problems here: most likely the regexp should be adjusted, but also the syntax with [] should be documented in the man page.

Yeah, we need to adjust the test, and update the man page. Note that the syntax is handled by the following: https://github.com/systemd/systemd/blob/6b000af4f206a87f424f05c163ea818b142e372e/src/libudev/libudev-util.c#L22

**Comment by khfeng** [comment]
> > DMI chassis types are quite often incorrectly filled. So I believe we should stick to current model, i.e. each model has its own quirk entry.
> > @jwrdegoede may have more input on this.
> 
> On Dell's systems you've seen it wrong? Any details would be really useful. To be honest - I'd be really surprised it's wrong on any major OEM - particularly on systems shipping with Windows 10. The specific requirements for https://docs.microsoft.com/en-us/windows-hardware/test/hlk/testref/751627d6-dde1-47d9-81a4-057a2aa1a3be prohibit it being wrong.

No I don't think I've seen wrong chassis type for major OEMs. Thanks for your confirmation.

File: rules.d/60-sensor.rules (hunk 2 above)
Comment (khfeng) [reply rules.d/60-sensor.rules:7]: HID-SENSOR-* is a platform driver created by ISH,  so please remove test on usb and i2c.

Comment (KaiChuan-Hsieh) [reply rules.d/60-sensor.rules:7]: Thanks for reviewing it, done.

Comment (khfeng) [reply rules.d/60-sensor.rules:7]: LGTM.

**Comment by jwrdegoede** [comment]
> > DMI chassis types are quite often incorrectly filled. So I believe we should stick to current model, i.e. each model has its own quirk entry.
> > @jwrdegoede may have more input on this.
> 
> On Dell's systems you've seen it wrong? Any details would be really useful. To be honest - I'd be really surprised it's wrong on any major OEM - particularly on systems shipping with Windows 10. The specific requirements for https://docs.microsoft.com/en-us/windows-hardware/test/hlk/testref/751627d6-dde1-47d9-81a4-057a2aa1a3be prohibit it being wrong.

Sorry, but trusting the chassis-type is not a good idea. I'm afraid that major OEMs, including Dell, do **often** get it wrong. Here are a couple of examples, which took me like 1 minute to find (copy and pasting the info was more work then finding it) :

1. XPS 13 7390 2-in-1:
    https://github.com/linuxhw/DMI/blob/master/Notebook/Dell/XPS/XPS%2013%207390/7B08B2130485

```
BIOS Information
       Release Date: 11/08/2019
...
Chassis Information
        Manufacturer: Dell Inc.
        Type: Notebook
```

2. Venue 8 pro 5855 tablet, from my personal hw collection: chassis_type: 8 ("Portable")

3. Venue 10 Pro 5055 tablet:
    https://github.com/linuxhw/DMI/blob/master/Notebook/Dell/Venue/Venue%2010%20Pro%205055/FCB31C87135A

```
BIOS Information
	Release Date: 04/24/2018
...
Chassis Information
	Manufacturer: Dell Inc.
	Type: Portable

```

4. Venue 11 Pro 7130 tablet from my personal hw collection: chassis_type: 8 ("Portable")

**Comment by KaiChuan-Hsieh** [comment]
@superm1 
Would you help to check @jwrdegoede 's finding?

**Comment by jwrdegoede** [comment]
So looking at the udev rule changes here, there seem to be several issues with them.

1. They skip the entire hwdb import for any HID sensors for any OEM when the chassis-type is 30|31|32

I assume the intention is to skip the following hwdb match: "sensor:modalias:platform:HID-SENSOR-200073:dmi:*svnDell*" on devices with these chassis-types so as to not set "ACCEL_LOCATION=base" there.

But the net result is that **any** hwdb imports on **any** HID sensors for **any** OEM are skipped when the chassis-type is 30|31|32. This disables the option to apply an ACCEL_MOUNT_MATRIX to HID sensors on any devices with a chassis-type of 30|31|32.

2. The commit message says "For product's chassis_type is not {30, 31, 32}, disable rotation on all Dell products with Intel sensor_hub embedded." beside this doing much more (see 1.) the "Intel sensor_hub" part is also not correct. The HID-SENSOR-200073 matches *any* accelerometer using the HID protocol including many non Intel HID sensor-hubs connected over i2c, such as the SMO91D0 HID sensor hub used on the Dell Venue 8 pro 5855 and on the Dell Venue 11 Pro 7130. So this part of the udev rule also will trigger (evaluate to true) in many cases where it should not trigger.

Note that the chassis-type not being reliable means that at least on both Dell Venue-s above the udev rule to skip the import (which has issues of its own) will not trigger, causing the hwdb import to proceed normally which will now trigger on:

```
sensor:modalias:platform:HID-SENSOR-200073:dmi:*svnDell*
    ACCEL_LOCATION=base
```

Causing rotation to stop working on both Dell Venue tablets (and likely on many other Dell models too).

Long story short, this pull-req is simply no good, so this gets a big an loud NACK from me.

The udev rule matches in way too much cases, while at the same time not matching in some cases where it should the hwdb import to get skipped. More in general trying to skip the import like this just is a very bad idea, I hope my comments explain why this is bad idea.

And the hwdb match will then lead to disabling display rotation on many devices where it used to work.

**Comment by jwrdegoede** [comment]
> @superm1
> Would you help to check @jwrdegoede 's finding?

I'm not sure what you expect @superm1 to check here. This is not some opinion of mine, this is actual info coming from actual hardware showing that within minutes I can find not 1 but 4 examples that the chassis-type can NOT be relied on in this manner.

More over as I explained in my second comment this pull-req also has all kinds of other issues.

Sorry of this sounds harsh, but this pull-req simply needs to be closed as it is no good and then you need to come up with another approach for the problem which you are trying to solve.

Please close this pull-req and when you submit a new pull-req with another approach at a @jwrdegoede comment there, so that I get notified and that I can review the new pull-req.

**Comment by KaiChuan-Hsieh** [comment]
@jwrdegoede 

thanks for your comment and review. close the pull-req, since it causes many issues.

**Comment by superm1** [comment]
It's unfortunate to hear several examples of mistakes. The venue tablets predated windows 10 so I would give them a pass, but I'll need to talk to the windows team about the situation with the recent 2-in-1. Being provided evidence of the problems I agree this is the wrong direction to decide based upon.

**Comment by jwrdegoede** [comment]
> It's unfortunate to hear several examples of mistakes. The venue tablets predated windows 10 so I would give them a pass, but I'll need to talk to the windows team about the situation with the recent 2-in-1. Being provided evidence of the problems I agree this is the wrong direction to decide based upon.

This discussion, as well as the SW_TABLET_MODE which you and I had elsewhere very recently, makes me believe that Windows has some way other then the chassis-type to figure out if something is a 2-in-1 or a regular clamshell model.

It seems that when a model is not a 2-in-1 windows ignores any accelerometers (no display rotation) and I guess also the tablet-mode switch info. The question is if Windows does not use the chassis-type for this then what does it use ? If you can figure that out somehow then that would be great.

**Comment by superm1** [comment]
@jwrdegoede I had some discussion with the Windows team in your examples.  The tablets that you linked pre-dated SMBIOS 3.0, and so the "Tablet", "Convertible", and "Detachable" chassis types didn't yet exist.

Regarding the more recent example of 7390-2-in-1, that's indeed a mistake and I asked them the impact of such a mistake, does this cause problems in the OS because of it relying upon that field?  They told me it doesn't cause functional Win10 problems but telemetry problems.  I would expect that should be fixed in a firmware upgrade and i'll raise it with that team.

> It seems that when a model is not a 2-in-1 windows ignores any accelerometers (no display rotation) and I guess also the tablet-mode switch info. The question is if Windows does not use the chassis-type for this then what does it use ? If you can figure that out somehow then that would be great.

I was told it's entirely handled by the "Intel chipset driver".  It sets initial state with the OS, and sends events to switch modes.  The OS only offers rotation when running in "tablet" mode.  To me, that means we should probably put a lot more weight in Linux on the intel-vbtn driver and intel-hid drivers.  We should be using the presence of switches provided by intel-vbtn to drive decisions in userspace.

**Comment by jwrdegoede** [comment]
> I was told it's entirely handled by the "Intel chipset driver". It sets initial state with the OS, and sends events to switch modes.

Ok, so we are back to needing to poke Intel to get more info how their driver for this works.

> The OS only offers rotation when running in "tablet" mode. To me, that means we should probably put a lot more weight in Linux on the intel-vbtn driver and intel-hid drivers. We should be using the presence of switches provided by intel-vbtn to drive decisions in userspace.

I somewhat agree starting with Fedora-33 / GNOME-3.38 GNOME will disable rotation if SW_TABLET_MODE is being reported by an evdev and that evdev is currently reporting SW_TABLET_MODE=0.

But there are many tablets and 2-in-1s which do not use the intel-vbtn driver and intel-hid drivers at all. Some of these might still advertise SW_TABLET_MODE through some other means. But some do not advertise it all, so we should only disable rotation when an evdev advertising SW_TABLET_MODE is present and it is currently advertising SW_TABLET_MODE=0, which as mentioned is a change which GNOME3 has already implemented.

I would not be surprised if the 2 models from #17281 which AFAIK are the reason to start this pull-req:

1. Dell Latitude 7420
2. Dell Latitude 9520

Actually do implement the intel-vbtn interface, but we are currently ignoring the SW_TABLET_MODE advertised by the intel-vbtn interface on them because of the chassis-type check in the Linux intel-vbtn driver to avoid issues with us falsely reporting SW_TABLET_MODE=1 on many normal clamshall laptops.

Which brings us back to the discussion we had elsewhere about how we really need more info from Intel to properly fix the intel-vbtn driver...

@KaiChuan-Hsieh can you perhaps provide acpidumps of the Dell Latitude 7420 and Dell Latitude 9520? I would like to check if they implement the intel-vbtn ACPI interface and what their implementation looks like. You mentioned that the 9520 comes in both clamshell and convertible format, can you provide the acpidumps for both ?

**Comment by superm1** [comment]
>But there are many tablets and 2-in-1s which do not use the intel-vbtn driver and intel-hid drivers at all. Some of these might still advertise SW_TABLET_MODE through some other means. But some do not advertise it all, so we should only disable rotation when an evdev advertising SW_TABLET_MODE is present and it is currently advertising SW_TABLET_MODE=0, which as mentioned is a change which GNOME3 has already implemented.

This is true.  I was told that before we had/used Intel virtual buttons driver we did it with a Dell ACPI interface (IIRC the same one used for `dell-rbtn`).  Can you enumerate some examples that do rotation now but don't have a `SW_TABLET_MODE`?

**Comment by jwrdegoede** [comment]
> Can you enumerate some examples that do rotation now but don't have a `SW_TABLET_MODE`?

How long a list do you want ? From the top of my head here are a few examples:

1. Almost all earlier Bay Trail based 2-in-1s (detachable kbd models) lack any way to report SW_TABLET_MODE, I can make a long list of the models which I know of, which will just be the tip of the iceberg
2. All Asus Transformer models use an Asus specific WMI interface to report SW_TABLET_MODE. This WMI interface is available on non Transformers too and always reports SW_TABLET_MODE=1 there, so we have an allowlist with Transformor models on which we use this WMI API, any models missing from the allowlist will be missing SW_TABLET_MODE reporting
3. Asus yoga style 2-in-1s, same story as Transformers, but different WMI API, currently the allowlist has only 1 model, so almost all Asus yoga style 2-in-1s miss this.
4. Various HP yoga style 2-in-1s miss this, they use the intel-vbtn driver but it has issues so they are not on the allowlist inside the intel-vbtn driver
5. Many budget brand yoga style 2-in-1s use 2 accelerometers instead of having a switch do detect being flipped into tablet mode in their hinges (I guess the switch is expensive to implement). There is some discussion here:
https://gitlab.freedesktop.org/hadess/iio-sensor-proxy/-/issues/216 about supporting the 2 accelerometer setup in userspace but for now this means no SW_TABLET_MODE reporting, Examples of this are the popular Tectlast F5, BMAX Y11 and BMAX Y13 models (AFAIK) + various Medion and Trekstor models
6. The Medion Akoya E2215T yoga style 2-in-1 has no way to detect SW_TABLET_MODE at all, instead it has a physical switch on the side to cut the power to the builtin USB kbd+touchpad combo to avoid spurious input while in tablet mode (yes really)

TL;DR: when there is a SW_TABLET_MODE we should be able to rely on it to disable rotation when it reports 0, but we cannot infer any meaning from the lack of SW_TABLET_MODE reporting.

**Comment by superm1** [comment]
> instead it has a physical switch on the side to cut the power to the builtin USB kbd+touchpad combo to avoid spurious input while in tablet mode (yes really)

Wow.

> TL;DR: when there is a SW_TABLET_MODE we should be able to rely on it to disable rotation when it reports 0, but we cannot infer any meaning from the lack of SW_TABLET_MODE reporting.


OK, I see what you mean.  If we can get a complete enough list do you think we could get to a point that perhaps there is an allowlist of things that should do _rotation_ and instead rely upon either SW_TABLET_MODE or that list?  I think it's probably better from userspace to not have everything rotating around, but rather work towards making sure that switch shows up on new stuff and older stuff that it can't show up for some reason but should still rotate it's on a list to treat it that way.

**Comment by jwrdegoede** [comment]
> OK, I see what you mean. If we can get a complete enough list do you think we could get to a point that perhaps there is an allowlist of things that should do _rotation_ and instead rely upon either SW_TABLET_MODE or that list?

I'm afraid that such an allowlist will never be anywhere near complete. There is an almost infinite amount of Bay Trail, Cherry Trail, Apollo Lake and Gemini Lake basec cheap 2-in-1s models out there and new ones are still being introduced. In the Netherlands (with all its 17 million people) alone I know of 3 companies who have bought such models from ODM's in Chenzhen and slapped on their own label including adding their own DMI strings.

So I don't think a generic allow-list is a good idea. But we could do e.g. a Dell based allow-list + chassis-type approach using udev + hwdb, to at least solve this problem in a generic manner for Dell and maybe some other major brands too.

Note we already something similar for HP models to avoid the HDD-fall-protection sensor HP uses to cause undesirable rotation on many HP models we have these lines in 60-sensors.hwdb:

# Most HP Laptop using the lis3lv02d device have it in the base,
# mark these sensors as such.
sensor:modalias:platform:lis3lv02d:dmi:*svn*Hewlett-Packard*:*
sensor:modalias:platform:lis3lv02d:dmi:*svn*HP*:*
 ACCEL_LOCATION=base
